src/vm_cpu.rs

```diff
@@ -1102,63 +1098,69 @@ pub enum VmCpuStatus {
 struct VmCpusInner {
     // Locking: status must be locked before vcpu.
     status: RwLock<VmCpuStatus>,
-    vcpu: Mutex<VmCpu>,
+    vcpu: Once<PageBox<Mutex<VmCpu>>>,
+}
+
+impl Default for VmCpusInner {
+    fn default() -> Self {
+        VmCpusInner {
+            status: RwLock::new(VmCpuStatus::NotPresent),
+            vcpu: Once::new(),
+        }
+    }
 }
 
 /// The set of vCPUs in a VM.
 pub struct VmCpus {
-    inner: PageVec<VmCpusInner>,
+    num: RwLock<usize>,
+    vcpus: [VmCpusInner; MAX_CPUS],
 }
 
 impl VmCpus {
     /// Creates a new vCPU tracking structure backed by `pages`.
-    pub fn new(
-        guest_id: PageOwnerId,
-        pages: SequentialPages<InternalClean>,
-        page_tracker: PageTracker,
-    ) -> Result<Self> {
-        let num_vcpus = pages.length_bytes() / VM_CPU_BYTES;
-        if num_vcpus == 0 {
-            return Err(Error::InsufficientVmCpuStorage);
-        }
-        let mut inner = PageVec::new(pages, page_tracker);
-        for i in 0..num_vcpus {
-            let entry = VmCpusInner {
-                status: RwLock::new(VmCpuStatus::NotPresent),
-                vcpu: Mutex::new(VmCpu::new(i, guest_id)),
-            };
-            inner.push(entry);
+    pub fn new() -> Self {
+        let vcpus = [(); MAX_CPUS].map(|_| VmCpusInner::default());
+        Self {
+            num: RwLock::new(0),
+            vcpus,
         }
-        Ok(Self { inner })
     }
 
     /// Returns the number of vCPUs in this `VmCpus`.
     pub fn num_vcpus(&self) -> usize {
-        self.inner.len()
+        *self.num.read()
```

Comment:
Does this re-define `num_vcpus` from number of possible to number added? Does that break `validate_imsic_addrs`?

Reply:
Oh yes. In my defense a name such as `max_vcpus()` would be more appropriate. This returns, in fact, the number of CPUs. Permission to change it to `max_vcpus()` requested.

Reply:
sgtm, thanks!

However, if you modify the storage and allow looping over only present vCPUs, it might not be needed.